lib/protobuf/field/field_array.rb

```diff
@@ -21,21 +21,51 @@ def initialize(field)
       #
 
       def []=(nth, val)
-        super(nth, normalize(val)) unless val.nil?
+        if nth < 0 || nth > size
+          fail IndexError, "Index #{nth} is outside valid range 0...#{size}"
+        end
+        super(nth, normalize(val))
       end
 
       def <<(val)
-        super(normalize(val)) unless val.nil?
+        super(normalize(val))
       end
 
-      def push(val)
-        super(normalize(val)) unless val.nil?
+      alias_method :push, :<<
+
+      def unshift(val)
+        super(normalize(val))
       end
 
       def replace(val)
         raise_type_error(val) unless val.is_a?(Array)
-        val.map! { |v| normalize(v) }
-        super(val)
+        super(val.map { |v| normalize(v) })
+      end
+
+      def map!(&block)
+        if block_given?
+          new_ary = map(&block)
+          replace(new_ary)
+        else
+          super()
+        end
+      end
+
+      def fill(*args, &block)
```

Comment:
I might be wrong, but I think that `fill` and `insert` can add nil values when they go out of bounds, e.g. 
```
 > t = [1,2,3]
 => [1, 2, 3] 
 > t.fill(4, 8, 2)
 => [1, 2, 3, nil, nil, nil, nil, nil, 4, 4] 
```

I think your methods might suffer the same problem? If I'm wrong, there should still probably be IndexError tests for this field_array_spec.